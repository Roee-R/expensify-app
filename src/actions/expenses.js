import uuid from 'uuid'; // generate id

 // Add expense 
 export const addExpense = ( 
    { 
        description='', note='', amount=0, createdAt=0 // USING OBJECT DESTRUCTURIN
    } ={})=>(
        { // return OBJECT
            type: 'ADD_EXPENSE',
            expenses: {
                id: uuid(),
                description,
                note,
                amount,
                createdAt
            }
        }   
    )

// Remove expanse
export const removeExpense = ({id})=>(
        { // return OBJECT
            type: 'REMOVE_EXPENSE',
            id
        }   
    )
    
// Edit expanse

export const editExpense = (id, updates) =>({
    type: 'EDIT_EXPENSE',
    id,
    updates
})