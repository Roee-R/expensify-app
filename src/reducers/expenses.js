// expenses Reduser

const expensesReduserDefaultState = []; // default array state

// Create expensesReduser Reducer
const expensesReduser = (state = expensesReduserDefaultState, action)=>{ // action is the extra propeties that we delivar via reducer object TYPES functions like editExpense
    switch(action.type){
        case 'ADD_EXPENSE':{
            return [...state, action.expenses]; // ...state is new syntax for arrays "spearign array"
        }
        case 'REMOVE_EXPENSE':{
            return [...state].filter(( {id} )=>id!==action.id);
        }
        case 'EDIT_EXPENSE':{
            return state.map(( expense )=>{
                if(expense.id===action.id){
                    return {
                        ...expense, ...action.updates //{...expense} is new syntax for arrays "spearign object"
                    };
                }else{
                    return expense
                }
            })
        }
        default:{
            return state;
        }
    }
}

export default expensesReduser;