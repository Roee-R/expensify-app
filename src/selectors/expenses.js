import moment from 'moment'

// sort and filter function
const getVisibleExpense = (expenses,{text, sortBy, startDate, endDate})=>{ // destructuring

    return expenses.filter((expense)=>{ // filter return just values with number bitween start date andend date
            const momentCreatedAt= moment(expense.createdAt)
            const startDateMatch=startDate?startDate.isSameOrBefore(momentCreatedAt, 'day'):true
            const endDateMatch=endDate?momentCreatedAt.isSameOrBefore(endDate, 'day'):true   
            const textMatch = expense.description.toLowerCase().includes(text.toLowerCase()); 

            return startDateMatch && endDateMatch && textMatch;
        }).sort((a,b)=>{ // sorting by amount or by date
            if(sortBy.toLowerCase()==='amount')
                return b.amount-a.amount;
            else if(sortBy.toLowerCase()==='date'){
                return b.createdAt-a.createdAt; 
            }   
        })  
}

export default getVisibleExpense;