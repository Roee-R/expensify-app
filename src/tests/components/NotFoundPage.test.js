import React from 'react';

import expenses from '../fixtures/expenses';
import {shallow} from 'enzyme';
import NotFoundPage from '../../components/HelpPage';

describe('Not found page test', ()=>{
    test('should render not found page', ()=>{
        const wraper = shallow(<NotFoundPage />)
        expect(wraper).toMatchSnapshot();
    })
})
