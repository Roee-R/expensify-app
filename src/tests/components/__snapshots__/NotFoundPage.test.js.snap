// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Not found page test should render not found page 1`] = `
<div>
  Wellcome from my help page component
</div>
`;
